pytket/extensions/qiskit/qiskit_convert.py

```diff
@@ -247,43 +248,157 @@ def _qpo_from_peg(peg: PauliEvolutionGate, qubits: list[Qubit]) -> QubitPauliOpe
+            return OpType.CnZ
+        case _:
+            if (
+                c_gate.base_gate.base_class in _known_qiskit_gate
+                or c_gate.base_gate.base_class is UnitaryGate
+            ):
+                return OpType.QControlBox
+            else:
+                raise NotImplementedError(
+                    "Conversion of qiskit ControlledGate with base gate "
+                    + f"base gate {c_gate.base_gate}"
+                    + "not implemented."
+                )
+
+
+def _optype_from_qiskit_instruction(instruction: Instruction) -> OpType:
+    """Get a pytket OpType from a qiskit Instruction."""
+    if isinstance(instruction, ControlledGate):
+        return _get_controlled_tket_optype(instruction)
+    try:
+        optype = _known_qiskit_gate[instruction.base_class]
+        return optype
+    except KeyError:
+        raise NotImplementedError(
+            f"Conversion of qiskit's {instruction.name} instruction is "
+            + "currently unsupported by qiskit_to_tk. Consider "
+            + "using QuantumCircuit.decompose() before attempting "
+            + "conversion."
+        )
+
+
+def _add_state_preparation(
+    tkc: Circuit, qubits: list[Qubit], prep: Initialize | StatePreparation
+) -> None:
+    """Handles different cases of Initialize and StatePreparation
+    and appends the appropriate state preparation to a Circuit instance."""
+
+    # Check how Initialize or StatePrep is constructed
+    # With a string, an int or an array of amplitudes
+    if len(prep.params) != 1:
+        if isinstance(prep.params[0], str):
+            # Parse string to get the right single qubit gates
+            circuit_string: str = "".join(prep.params)
+            circuit = _string_to_circuit(
+                circuit_string, prep.num_qubits, qiskit_prep=prep
+            )
+            tkc.add_circuit(circuit, qubits)
         else:
-            raise ValueError(
-                f"Cannot parse string for character {char}. "
-                + "The supported characters are {'0', '1', '+', '-', 'r', 'l'}."
+            amplitude_array: NDArray[np.complex128] = np.array(prep.params)
+            pytket_state_prep_box = StatePreparationBox(
+                amplitude_array, with_initial_reset=(type(prep) is Initialize)
             )
 
-    return circ
+            # Need to reverse qubits here (endian-ness)
+            reversed_qubits = list(reversed(qubits))
+            tkc.add_gate(pytket_state_prep_box, reversed_qubits)
+    else:
+        if isinstance(prep.params[0], complex):
+            # convert int to a binary string and apply X for |1>
+            integer_parameter = int(prep.params[0].real)
+            bit_string = bin(integer_parameter)[2:]
+            circuit = _string_to_circuit(bit_string, prep.num_qubits, qiskit_prep=prep)
+            tkc.add_circuit(circuit, qubits)
+        else:
+            raise TypeError(
+                "Unrecognised type of Instruction.params "
```

Comment:
I'm not so satisfied with this `_add_state_preparataion` function. There are 3 different ways to create a `StatePreparation` or `Initialize` object. Couldn't simplify the handling all that much.